common/flatpak-instance.c

```diff
@@ -313,6 +317,47 @@ flatpak_instance_get_info (FlatpakInstance *self)
   return priv->info;
 }
 
+GStrv
+flatpak_instance_get_run_environ (FlatpakInstance *self)
+{
+  FlatpakInstancePrivate *priv = flatpak_instance_get_instance_private (self);
+
+  return priv->run_environ;
+}
+
+static GStrv
+get_run_environ (const char *dir)
+{
+  g_autofree char *file = NULL;
+  g_autofree char *environ_data = NULL;
+  gsize environ_size;
+  g_autoptr(GError) error = NULL;
+  g_autoptr(GPtrArray) env_vars = NULL;
+  int i;
+  gsize env_var_size;
+
+  file = g_build_filename (dir, "run-environ", NULL);
+
+  if (!g_file_get_contents (file, &environ_data, &environ_size, &error))
+    {
+      g_info ("Failed to load run-environ file '%s': %s", file, error->message);
+      return NULL;
+    }
+
+  env_vars = g_ptr_array_new_with_free_func (g_free);
+
+  for (i = 0; i < environ_size - 1; i += env_var_size)
+    {
+      env_var_size = strlen (&environ_data[i]) + 1;
+
+      g_ptr_array_add (env_vars, g_memdup2 (&environ_data[i], env_var_size));
```

Comment:
If the variable is syntactically invalid (no `=` sign, like `GOOD=good\0BAD\0`, or `=` sign at position 0, like `GOOD=good\0=bad\0`), then I think we need to detect that. In this case it seems more appropriate to log (perhaps at info level) and exclude that one from the array, rather than raising an error.

Reply:
I was actually thinking whether we should treat the contents of the environment file as potentially malicious and handle its parsing more robustly, or consider this to be a "trusted file" and allow ourselves to make assumptions about its correct formatting.

To explain this further: with the assumption that our code would always produce correct `env -0` formatting for the file (unless we have a bug that would need to be fixed anyway), and with the assumption that we shouldn't try to handle filesystem corruption gracefully, the only case where one could expect the file format to be non-compliant is if a malicious actor had modified it.

My thinking was that if a malicious actor has write access to this file, we are already compromised and there's little we can do to prevent DoS attacks (simply delete the file or make it 100GB) or more elaborate attacks (inject environment variables to app sandboxes).

I don't mind being protective its parsing, but I preferred to avoid the added complexity without a good reason. There can be many attack vectors that we'd need to handle if we were to consider the contents of this file untrusted.

With that in mind, do you still prefer to make this parsing more robust? If so, how much more robust? Would sharing code with `flatpak_context_parse_env_block` suffice?

As a side-note, I prefer to abort with an error in case of bad formatting than silently ignore environment variables and result in slightly broken behavior. I also prefer to treat logging as "silent" because simple users would never see the log or know to look for it, but would still experience slightly broken behavior.

Reply:
> My thinking was that if a malicious actor has write access to this file, we are already compromised

Oh, yeah, definitely. This is robustness rather than security (and so is `--env-fd`, for that matter) - defending against mistakes rather than attackers, with validation at the boundary to make it legitimate for other modules to make reasonable assumptions and be confident that a crash won't result.

The main reason why I was careful about validation when I added `--env-fd` is for the same reason that new syscalls are meant to fail if `flags != 0`. I wondered whether `FOO=foo\0BAR\0` should mean "set FOO to foo, and then unset BAR". I decided against that in the end, but I wanted to give myself the option to introduce that behaviour later, and I wouldn't be able to do that without a compatibility break if it was something that already worked but had a different result.

> As a side-note, I prefer to abort with an error in case of bad formatting than silently ignore environment variables and result in slightly broken behavior

That's fair. I'd be OK with having this fail recoverably, but that would require passing a GError up to a layer where failing is reasonable to do. However, it is not OK for this to `abort()` or `g_error()`, because `FlatpakInstance` is public API in libflatpak (via `flatpak_instance_get_all()`). For example, it would not be OK for `flatpak ps` to crash just because there's an app running that was started before upgrading Flatpak.

It's a bit awkward that this is a private method on a public object. We're constrained by `FlatpakInstance` not being `GInitable` (and retroactively making it `GInitable` would arguably be an incompatible change), which means its constructor can't fail; but if we load the environment block in the constructor, then that means failure to load the environment can't be signalled properly.

Can we, instead, make this

```
const char * const *flatpak_instance_get_run_environ (FlatpakInstance *self, GError **error);
```

with the environment block loaded on-demand and cached in `self` after the first successful load; or even

```
GStrv flatpak_instance_get_run_environ (FlatpakInstance *self, GError **error);
```

with it loaded every time (no difference, in practice)? That way, if loading the environment block fails, the portal can do something like (pseudocode)

```
if (flatpak_instance_get_run_environ () succeeds)
  use its result
else if (g_error_matches (error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND))
  g_info(), use g_get_environ ()
else
  raise the error, return failure
```

Reply:
> Would sharing code with flatpak_context_parse_env_block suffice?

Yes.

Reply:
> However, it is not OK for this to `abort()` or `g_error()` ...

Sorry for the poor choice of words, I did not actually mean "abort" in that sense. I meant return from the function early with a `NULL` pointer, rather than return a partially-loaded environment. I don't see how returning a partially-loaded environment could be useful or desirable.

I'm still not happy with the idea of tolerating a missing/non-conformat environment file as a general rule, but I also don't see a better way to approach this if we intend to achieve both forward and backwards compatibility between `flatpak run` and `flatpak-portal`.

(For posterity: I am a strong advocate of implementing logic as atomic operations at every level of the stack; if a function fails to perform its operation in full it should indicate this to the caller and have zero side-effects. Allowing partial state may give the false impression of robustness, but in fact results in unpredicable behavior. I mentioned this as "slightly broken" behavior in a few comments. Tolerating partial state at scale eventually leads to borderline-undefined behavior; the slightest abnormal operation turns into a butterfly effect and the relation between the symptoms of an issue and its root cause appears entirely circumstantial.)

Reply:
I added `flatpak_parse_env_block()`, and used it in `flatpak_context_parse_env_block()` and `get_run_environ()`. This means various format errors in the environment block are now detected and reported.

I split the comment about the handling of errors in `get_run_environ()`/`flatpak_instance_get_run_environ()` to its own thread. I'll resolve this one.